src/main/scala/org/deepdive/ddlog/DeepDiveLogSemanticChecker.scala

```diff
@@ -142,6 +149,94 @@ object DeepDiveLogSemanticChecker extends DeepDiveLogHandler {
     }
   }
 
+  // collect variables used in the expression
+  def collectVariables(expr: Expr) : Set[String] = {
+    var set = new HashSet[String]()
+    def collectVariablesInner(e: Expr) : Unit = e match {
+      case VarExpr(name) => set += name
+      case FuncExpr(function, args, agg) => args foreach (collectVariablesInner(_))
+      case BinaryOpExpr(lhs, op, rhs) => {
+        collectVariablesInner(lhs)
+        collectVariablesInner(rhs)
+      }
+      case TypecastExpr(lhs, rhs) => collectVariablesInner(lhs)
+      case _ =>
+    }
+    collectVariablesInner(expr)
+    set
+  }
+
+  def collectVariables(pattern: Pattern, definitionOnly: Boolean) : Set[String] = {
+    var set = new HashSet[String]()
+    pattern match {
+      case VarPattern(name) => set += name
+      case ExprPattern(e) => if (!definitionOnly) set ++= collectVariables(e)
+      case PlaceholderPattern() =>
+    }
+    set
+  }
+
+  def collectVariables(cond: Cond) : Set[String] = {
+    var set = new HashSet[String]()
+    cond match {
+      case ComparisonCond(lhs, op, rhs) => {
+        set ++= collectVariables(lhs)
+        set ++= collectVariables(rhs)
+      }
+      case CompoundCond(lhs, op, rhs) => {
+        set ++= collectVariables(lhs)
+        set ++= collectVariables(rhs)
+      }
+      case NegationCond(c) => set ++= collectVariables(c)
+    }
+    set
+  }
+
+  // collect variable
+  def collectVariablesFromBody(body: Body, definitionOnly: Boolean) : Set[String] = {
+    var set = new HashSet[String]()
```

Comment:
You don't really need this accumulator var here, because you can directly return the set from all cases.